### ISSUE TITLE
End support for PowerShell v3/v4 in VSCode-PowerShell v2.0.0

### ISSUE THREAD

**Issue by rjmholt** [PowerShell/vscode-powershell#1310]
Currently, supporting PowerShell v3 and v4 creates several challenges for us:

* To get PSReadLine into the extension, we need changes coming in PSReadLine 2.0, which does not support v3/v4.

* We are forced to use the v3 AST to support v3, which means:
    - [No support for classes or enums](https://github.com/PowerShell/PowerShellEditorServices/issues/14)
    - [Unable to use modern AST APIs](https://github.com/PowerShell/PowerShellEditorServices/issues/276)

* Incompatibility in the APIs between v3, v4 and v5 mean that there are bugs that are hard to fix (and even hard to reproduce, since most of us don't have access to a machine running PowerShell v3 or v4).

* Of the two CI systems we currently use and the others we are looking to invest in, none supports PowerShell v3/v4 natively.

* v3/v4 usage is relatively low, and decreasing. With Windows Server 2019 around the corner and Server 2012 mainstream EOL also coming up this year, even more users will be moving to a newer PowerShell version.

* The code base is fragmented and hard to maintain because:
    - We have to compile against .NET Framework 4.5.1 **and** .NET Standard 1.6, when we would ideally compile against only .NET Standard. (This also spurs us toward better cross platform support)
    - We also have compile time PowerShell version targets.
    - Both of which lead to crazy .csproj files like [this one](https://github.com/PowerShell/PowerShellEditorServices/blob/master/src/PowerShellEditorServices/PowerShellEditorServices.csproj).
    - The .NET Framework/Standard stuff means the backend codebase is [riddled with `#ifdef`s](https://github.com/PowerShell/PowerShellEditorServices/blob/0cef8c0372864295ac87f3827295c2d18cefc2b6/src/PowerShellEditorServices.Host/EditorServicesHost.cs#L153).
    - There are also [`#ifdef`s for v3/v4 support](https://github.com/PowerShell/PowerShellEditorServices/blob/cb68635231ecd92412527e6e38306b1143ac6f73/src/PowerShellEditorServices/Session/PowerShell5Operations.cs#L14).
     - There's also version conditional logic in the build and run tooling, but that's easier to work with.

[This issue](https://github.com/PowerShell/PowerShellEditorServices/issues/558) contains most of the discussion that's already been had about dropping v3/v4.

My suggestion is:
- Include a deprecation warning for ~6 months, starting when v2.0.0 of the extension is released as stable.
- Actually drop support for v3 and v4 6 months after that date.
- Increment the major version number when we do it.
- Ensure that [PSScriptAnalyzer can analyze PowerShell feature support by version](https://github.com/PowerShell/PSScriptAnalyzer/issues/127) so that people can still write scripts for v3 and v4 using the VSCode extension (but with the extension host running v5+).

> ~Some additional things:~
> - ~An alternative to some of this is to refactor much of our code to abstract away the differences. However, that could be a costly effort and lead to significant regressions.~
> - ~We've also been talking about incrementing the major version of the extension with the addition of PSReadLine support. It seems a little cruel to me to drop v3/v4 support just as we add PSReadLine, but it also makes sense that both changes are big enough to mean a major version increase.~

### UPDATE

The addition of PSReadLine depends on a new version of PSReadLine that [requires PowerShell v5.1 or later](https://github.com/lzybkr/PSReadLine/blob/a878b3686cba23ad39fcd98a6b418c9db0bb77ab/PSReadLine/Changes.txt#L43). This is by far our biggest motivating factor for ending support for PowerShell v3/v4.

**Comment by bergmeister** [comment]
PSSA supports v3/4 using pragmas and conditional compilation. This way you can use the latest API is in the versions of PowerShell where they make sense, i.e. a rule analysing classes does not make sense on ps v4.
If this is too much pain, then you should consider having a v2 of the extension in parallel (with the pain of merging important commits back to v1) but dropping support is not ideal. I can understand that it would be nice for us developers but keep in mind that it is also the job of a developer to put up with that complexity to make life as easy as possible for the user.

**Comment by SeeminglyScience** [comment]
@bergmeister that's what we currently do, but @rjmholt did a great job at outlining the issues with that.  One of the biggest factors for me is the fact that we will only be able to support the latest version of PSReadLine, which requires PowerShell 5.1.  Adding to that is the sizable amount of refactoring required to get PSRL working (often requiring newer PowerShell API's). Meaning we'd be supporting two wildly different experiences and code bases.

Also I suspect that the vast majority of folks are either stuck on version 2, or they're running 5.1 with very little in between.  I'd be inclined to agree with you if we had the ability to support version 2, but we haven't since the beginning of the project (afiak).  Granted this is anecdotal and based mostly own my own experiences and discussions with others, but I would be incredibly surprised if versions 3/4 made up more than 5% of PSES usage between them.

**Comment by bergmeister** [comment]
> We are forced to use the v3 AST to support v3

What else do you expect? A user gets as many features as appropriate to the used version. By dropping support for v3 completely means that you change the feature coverage to 0%
>  Incompatibility in the APIs between v3, v4 and v5 mean that there are bugs that are hard to fix (and even hard to reproduce, since most of us don't have access to a machine running PowerShell v3 or v4).

Ok but it does not mean you have to fix all bugs in v3/4, you can still suggest users to upgrade should there be a blocking issue. Getting PowerShell v3/4 is as easy as creating a WinServer 2012/2012R2 VM, which takes less than 1 min to setup in Azure.

> v3/v4 usage is relatively low, and decreasing. With Windows Server 2019 around the corner and Server 2012 mainstream EOL around the corner, even more users will be moving to a newer PowerShell version.

Customers use OSs that are supported by MS and they still run v 2/3/4 out of the box and with updates. Most of my customers don't even know there is a newer version of PowerShell. Users then have to take what they get in a corporate environment as they do not have admin rights to install stuff on their machine.

> The code base is fragmented and hard to maintain 

As I said that is sometimes part of our job but of course there is a limit at some point...

Your last argument about PSReadline is more interesting because it would be a technical and unavoidable reason. I think we need to split the discussion/deprecation into 2 parts:
For the VSCode extension it might be OK to drop support because there is still the ISE but PSES is supposed to be a general purpose library for usage by any 3rd party (including non-Microsoft!). If I embedded an editor into my app (for easier inline editing) that used PSES then I need this to work on any ps version that is installed on the client machine without having to force them to upgrade since scripts can break with the upgrade!

**Comment by rjmholt** [comment]
I've opened this issue just so that we can have an open discussion about pros and cons where anyone can weigh in. This issue properly belongs in the [Editor Services repo](https://github.com/PowerShell/PowerShellEditorServices), but this repo has more visibility so seemed like the right place to start the discussion. I'll open an issue in PSES to link back to this one.

@bergmeister you make excellent points, and we should also consider the possibilities of forking PSES and refactoring it to handle old versions properly.

We do have a responsibility as developers, but our time is finite and we should plan accordingly -- time spent maintaining a more complex, ultra-backward compatible codebase will mean less features, bugfixes and performance improvements. As PowerShell develops and changes, we will have to support that too, and we're going to have to decide what gets less attention or is dropped (which could just translate to "we support everything, but not as well").

As an open source project, users are also contributors, and to my mind the responsibility is shared.

**Comment by SeeminglyScience** [comment]
> Ok but it does not mean you have to fix all bugs in v3/4, you can still suggest users to upgrade should there be a blocking issue. 

If that's acceptable, why not use an old version of the extension if you need support for v3/4?

> Getting PowerShell v3/4 is as easy as creating a WinServer 2012/2012R2 VM, which takes less than 1 min to setup in Azure.

That's a good solution for @rjmholt and @tylerl0706, but not for anyone else contributing to the project.

> Customers use OSs that are supported by MS and they still run v 2/3/4 out of the box and with updates. 

I have no doubt that tons of people are still running v2.  But we already don't and can't support that one.  What percentage are running 3/4 though?

> Your last argument about PSReadline is more interesting because it would be a technical and unavoidable reason. ... For the VSCode extension it might be OK to drop support

Keep in mind PSRL is *not* a VSCode feature.  The changes required are entirely within PSES and PSRL, any PSES enabled editor with a terminal will get the feature.

**Comment by tylerl0706** [comment]
Lots of great discussion here. I'm on the side of revving the major version and back-porting important bug fixes back to the v1.


Sorry for being terse - traveling :)

**Comment by vexx32** [comment]
While some % of people are still on v3/v4 of PowerShell, i think that in large proportion these are not on machines where anyone will be doing any regular developing. 

The ISE exists for one-off or occasional development work on those machines... But I think any serious PowerShell developer will be on 5.1, and many are starting to at least tinker with PS Core.

In other words, retaining overly complicated and time consuming support for these older platforms is unlikely to really benefit anyone we can maintain support for developing *for* these systems through version awareness (no classes when developing for PS v4, etc.), without needing to actually support development using these systems with the latest versions of these tools.

If you want similar functionality, there's nothing wrong with allowing people to still download and use the older versions that still support PowerShell versions 3 and 4, but I think it could be pretty hindering to both developers and users of these tools alike to unilaterally support as many older versions as we can.

It's a common problem: sure, we *can* spend exorbitant amounts of time and effort supporting older versions for as long as humanly possible. But I don't think we *should*.

**Comment by billkindle** [comment]
Still need time IMHO. Or at least give an option to switch to an older mode temporarily. Personally I have hoarded of systems that are on 2008 for another year or two. Yeah they are going to be replaced and yeah maybe still have ISE. I’m torn. I guess as long older versions would remain available for the fringe cases....

**Comment by Matalus** [comment]
I can understand the push to remove obsolete versions, but I'll make this argument, PowerShell tends to get used as a DevOps language, and at least in my case and I would wager many others. you don't always get to choose which version of PS  the customers you're working with are using. While most of what I do is in v5, I know I have customers that are still running v2 and they have zero plans to update ever.

**Comment by rkeithhill** [comment]
As has already been mentioned, the extension already does *not* support v2.

**Comment by dimitertodorov** [comment]
VSCode is mainly used on our development machines where it would be silly to run anything below 5.1

I develop a lot of Binary modules, and having C# interop work better would be great.
In particular, this issue would be great to close off.
PowerShell/PowerShellEditorServices#14


For testing, we always use remote machines for QA, and that is not dependent on having VSCode installed.

**Comment by DougPlumley** [comment]
I'm all for dropping it, our whole organization upgraded all servers to v5.

**Comment by AspenForester** [comment]
This system admin will not be above (mis)using this as an argument to management to get servers in the field upgraded. But in greater seriousness, I completely agree that development workstations should and tend to be running the current PowerShell.

Thank you for opening up this discussion!

**Comment by lehtoj** [comment]
All of my personal Windows machines at work have PowerShell 5.1. It is unfortunate fact that Microsoft has yet to update all their enterprise products to support PowerShell versions beyond 4.0, so that's what we still have to target when writing scripts.

Better support for 5.1, 6 and beyond would still be more than welcome. As long as dropping support for v3/v4 does not hinder our ability to target said versions too much, I say go for it.

**Comment by illegalOperation** [comment]
My clients that are still using legacy PowerShell are not using VScode.  I know there must be some edge cases where some servers can't be upgraded to current WMF (I'm looking at you old exchange) but I feel that if an IT shop is using VScode they are already on the ball on being current.  My clients that still rock older WMF are happy as a clam to use the ISE.

**Comment by lehtoj** [comment]
@ TheIncorrigible1 Well yes, I have yet to see a single Windows Server 2016 in the wild though. Support for v6 and beyond would be mostly for personal tinkering until a viable business case for it emerges.

**Comment by dragonwolf83** [comment]
This would be a different discussion had PS Editor Services been the framework for ISE. The ISE was and is the editor for PowerShell v2-5.1. This is a new framework and cannot be expected to just slide in perfect support to older versions of PowerShell that was not designed for the new functionality the framework supports.

Let's pose this question as if this was ISE. Has any version of ISE supported running older versions of PowerShell? AFAIK, you can only get the new ISE when you install the latest WMF. So had the ISE come out with some new uber feature, you wouldn't get it anyway unless you were on the latest.

Going forward, PowerShell is a side-by-side model. We have Windows PowerShell 5.1 and multiple versions of PowerShell Core v6.0 installed on the same box. PSES needs to maintain backwards compatibility support to 5.1 for the long-term. I think new features can still light up in newer versions, but not drop support for several years.

**Comment by tylerl0706** [comment]
Just to do a bit of moderation:

@dimitertodorov: Issue https://github.com/PowerShell/PowerShellEditorServices/issues/14 can be addressed quicker and with less pain if we drop v3/v4 🙂 

@TheIncorrigible1, @lehtoj value prop for PS 6 is not really a discussion point here but xplat is a benefit for those that manage a mix of linux/windows. Also it's _a lot_ faster. If there's something you're missing from PS Core, please [open an issue](https://github.com/powershell/powershell/issues) 😄

@dragonwolf83 the plan is that PSES will continue to support 5.1 for the long-term.

**Comment by TheIncorrigible1** [comment]
@dragonwolf83 Stop beating a dead horse.  ISE is deprecated and a different beast.

**Comment by tylerl0706** [comment]
Also to be clear, for those that still use v3/v4 you would still have 2 options:

* any v1.x PowerShell Editor Services editor like VSCode's extension (will still receive certain bug fixes)
* ISE which will be in Windows already (maintenance mode)

Also as folks have mentioned, you can still write a script that works on v3 while using a PSES editor (vscode) and 5.1 or 6.

In fact, PSScriptAnalyzer will have rules, I believe, that say "hey you're not compatible with this version" to help that scenario.

**Comment by dragonwolf83** [comment]
@TheIncorrigible1 by your definition, so is PowerShell v3-4. Besides, you missed the point. I'm not advocating ISE forever. I'm saying that Microsoft has no obligation to backport a new way to do an Editor to older versions. It is a nice to have but if it hinders development, I vote for the path @tylerl0706 mentioned above.

**Comment by SeeminglyScience** [comment]
@bergmeister 

> I am rather worried about the PSES dropping support rather than the VSCode extension.

I get the impression that you are worried this change is being made without other PSES implementations taken into consideration.  We may disagree on this change, but I'd like to reassure you that when it comes to PSES changes, my *primary* concern is how it could effect future implementations.  VSCode is a great example of what an implementation can do but the power of PSES comes from it's extensibility first and foremost.  It is with that in mind that I've been fighting for this change for some time.

> However, having said that, this would mean that it would be at least nice if both versions get installed when installing the extension and at runtime the extension decides which one to use, to avoid having to manually dig out the old one.

I'm not sure there is enough usage to justify implementing this.  Personally I think the same about forking/porting fixes but I think that's a reasonable enough compromise for those worried by the change.

> Keep in mind that people deliberately keep test boxes not updated to test that their PowerShell modules also work on older versions of PowerShell and therefore will need to visit the box from time to time for debugging purposes.

I think the big disconnect here, for me, is why VSCode (or any other editor) needs to be installed on these machines?  Wouldn't it be preferable to debug from a `PSSession`?  Or with something preexisting (ISE)?

> PSSA has a WMF4 image in the build matrix here to run tests against v4, which means that other people do not need any VM setup for a PR.

That's great for final checks before merge, but doesn't help as much with development.

> Why don't you just use the latest Parser of PSCore itself
> ...
>  This is possible because one of the NuGet packages of the powershell-core feed (I think it was either PowerShell.SDK or System.Automation.Management)

The SDK does contain all the code required for that, but in order to reference it you need to target `netcoreapp2.0`.  Right now we build two sets of assemblies, one for CoreCLR and one for FullCLR.  We can't reference the SDK in the FullCLR build which is where we would want to use a newer API, and we already reference it in the CoreCLR build.  Adding to that, PSES is hosted in a PowerShell session so even if we had a later version of SMA we could reference in FullCLR we can't override the one already loaded in the AppDomain.

I think there are significant enough reasons to look into ripping out the code for `SMAL.Parser` and adding some custom logic at some point.  But that doesn't solve this problem, a lot more than the parser has changed since 3.0.  Off the top of my head the debugger API has changed significantly from 3.0 to 5.1.

**Comment by bergmeister** [comment]
@seeminglyscience

> The SDK does contain all the code required for that, but in order to reference it you need to target netcoreapp2.0

This was true for version 6.0.1 but with 6.0.2 they did it properly using netstandard2.0, shouldn't this allow you then to use it in Windows PowerShell where it gets compiled for the fullclr? AppDomains should normally allow you to use a different version even when being hosted in Windows PowerShell.

We will probably never agree on PSES support but at least having some integrated legacy mode to keep basic functionality would be nice. I will ask management for details/clarification but I can imagine that if they get the choice between taking something that works on any version or taking something that only works on the latest version but receives great improvements, they are likely going to prefer the former option then. Our clients are unfortunately often in a big corporate environment that is not very flexible.
Personally I am all for using the latest and greatest stuff from a client and development point of view whereever possible but this is a tricky subject since it is affecting clients and not just us.

**Comment by SeeminglyScience** [comment]
> This was true for version 6.0.1 but with 6.0.2 they did it properly using netstandard2.0, shouldn't this allow you then to use it in Windows PowerShell where it gets compiled for the fullclr?

Maybe I'm getting it from the wrong place.  Is there a way to reference the SDK other than [this myget feed](https://powershell.myget.org/feed/powershell-core/package/nuget/Microsoft.PowerShell.SDK/6.0.2)?

**Comment by bergmeister** [comment]
Sorry, you are right about this one (I must have misremembered something from a different NuGet package). It would be interesting if there is a reason why it is netcore and not netstandard since the api set is not vastly different between the two (maybe it could be already netstandard and it was just published like that?)

Would it be too much of an impact on the footprint to publish it as self-contained then (i.e. include the required runtime)? Then you would have only one compilation target and could use the newer APIs in older versions of PS. There is already a feature (currently experimental so no official support at the moment) in the netcore linker to merge all published DLLs and trim out unneeded or duplicated IL (similar to how Mono already does it to have much smaller Xamarin apps) to end up with only very small binaries that need to be deployed?

**Comment by Juanito99** [comment]
Keep both versions for another few years. I would suggest until Windows 2012 is out of extended support.
There are servers, including those with Microsoft application on it which can't receive the required .net frameworks or wmf bases. 

Thanks :)

**Comment by pixelrebirth** [comment]
I completely support this in light of PSv5.1 Desktop being "Complete" as per Snover@Summit this year. The use cases to develop exclusively for <5.1 is difficult to justify in my opinion. 5.1 installs on win2008r2 for crying out loud and generally, your not running your developed code there anyhow, you are remoting in and doing a couple things and getting a result to be processed somewhere else.

I fully support this to improve vscode for 5.1.

**Comment by vexx32** [comment]
@Juanito99 Sure, there are servers running old PS versions. That's not what this is about, though. This is about dropping support for _developing_ on VS Code and other editors that utilise this framework on future versions, so that we can focus on implementing new features for the new versions, _such as_ providing better support for verifying a script will actually function on an earlier version of PS than is running on the development machine.

The sole reason I could see for not wanting to drop support for developing _in_ environments with earlier versions of PS is if you're trying to do dev work and writing most of your scripts on servers running old Windows versions.

I really wouldn't see the point of hamstringing better support for developing for those versions and future versions just to support a use case that doesn't even really make sense. If you absolutely have to fiddle with scripts in the environment itself using the machine you'll be testing on, you can still run an old version of the extensions or just run the ISE for that brief moment.

**Comment by Juanito99** [comment]
@vexx32 Thanks for explanation. It turns out that I misunderstood the intention. So if I will be not able to run VSCode incl. PS on legacy server it is absolutely acceptable for me. :+1: 

To me it is just important that I still can write scripts _for_ those old machines.

Please go ahead, then. You guys are really doing an amazing job! :smile:

**Comment by tylerl0706** [comment]
Just to clarify, @Juanito99, you will still be able to use VSCode, it just won't receive anymore features.

**Comment by Juanito99** [comment]
@tylerl0706 Nice, appreciate the clarification :bowing_man:

**Comment by Glober777** [comment]
I can't think of any scenarios in my environment, where anyone would need to run VSCode-PowerShell on machines with WMF older than 5.1, so I'm all for it, especially considering that dropping support for older WMFs can make PSReadLine support come out sooner.

On a side note, the idea about maintaining two PowerShell-VSCode extension versions looks interesting. What if instead of creating deprecation warnings, and so on, you just roll out a new extension (i.e. VSCode-PowerShell-new, I'm sure we can come up with a better title) that's not going to have a dependency on the older WMF versions? That way, there will be no need to wait on any deprecation dates and allow those that run newer WMF versions on their dev stations get new functionality much quicker.

**Comment by tylerl0706** [comment]
@Glober777 I was thinking more on the side of...

Currently we have just 

"PowerShell"

we will bring that forward to the v2 version of the PowerShell extension supporting 5.1+.

We'll add a 

"PowerShell (Legacy)"

For the 1.x version which will still work with v3/v4

**Comment by rjmholt** [comment]
I wasn't sure about this before, but worth noting now:

PSReadLine support depends on using PowerShell v5 and above -- the mechanism we need to integrate with the integrated console is in a newer PSReadLine version that does not support below PowerShell v5.

It's not necessarily true that adding PSReadLine would mean dropping v3/v4 support, but it would make things quite complex if we didn't.

**Comment by tylerl0706** [comment]
Just to add to Rob's comment, you can track the adding of PSReadline here:

https://github.com/PowerShell/vscode-powershell/issues/535

**Comment by uttampcu** [comment]
The vast majority on this page are all saying the same thing.  IDE != Runtime and it is therefore legitimate to have different requirements.  Give us the tools to help us develop our code.

**Comment by brwilkinson** [comment]
This seems like something that should have and RFC? Pretty significant decision/change.

"RFC documents for community feedback on design changes and improvements to PowerShell"

https://github.com/PowerShell/PowerShell-RFC/issues ?

I hope we can move forward with this change.

**Comment by joeyaiello** [comment]
Thanks for the feedback, everyone! This has been a super enlightening thread, and I was happy to hear folks chime in with their opinions at the last PowerShell Core Community Call as well.

@brwilkinson: the RFC process is intended for stuff within the PowerShell engine and language itself, this is really outside the scope, and I think we got the quality and quantity of feedback within this issue that we wanted. 

In the short-term, we should write a blog post about this to enumerate our thinking in a more public way, and give folks ~6 months for us to wind down support of what will likely remain as the "1.x" extension that continues to support PS 3.x/4.x without new features.

**Comment by swngdnz** [comment]
I can't keep up with all the issues!

It sounds like you've already made decisions, but I'll throw my 2 cents in here:

As long as I have to support old versions of Exchange, I need to be able to develop against old versions of PowerShell. Right now there are four versions of Exchange in play -- Exchange 2010, Exchange 2013, Exchange 2016, and Exchange 2019 (unreleased). So I develop against PS 3/4/5/5.1 and generally do it on those lab servers. Removing VSCode/PSES support for PS 3/4 will cause significant pain for me.

(And for those of you that are not aware, Exchange does not support you upgrading WMF versions. You use the version shipped with the OS where Exchange is installed.)

**Comment by vexx32** [comment]
You can still _use_ the old version(s) of the extension without the new features... this is more to break away from that so we can have new features on the versions of PS that support them.

**Comment by D3MaxT** [comment]
I also vote for only supporting 5.1 and Core with 2.x.

For older versions, use 1.x or use ISE or PowerShell Studio.

**Comment by DarthSpock** [comment]
I'd like to add my feedback here as well. I'm in a environment where I use a VM to do all my dev work and the only approved Windows VM is 7...So upgrading PS is impossible for me. If PS Core had better support for 7 then I could be in support of dropping it and using PS Core instead but since PS Core has no plans to support Win 7 due to it going EOL soon (which my org doesn't care about, they'll still use it anyways since it's just a VM and not the main box anyways.) I'm sure I'm not the only one in a similar situation so support is still needed but separating out the extension by whether you're on 10 or Core vice older is a good idea as long as support for both is still on going for at least the next 5 years. By then, my org may actually decide we need to upgrade to 10 (hopefully).

**Comment by vexx32** [comment]
You can upgrade Windows 7 machines to PowerShell 5.1. There's no need to suffer through on v2; it's _really_ not worth it. It's much more productive to use 5.1, there are a _ton_ of convenience features that make it a lot simpler to work with.

**Comment by swngdnz** [comment]
It may be _possible_ to update Win7 to PowerShell 5.1, but it may not be _allowed_.

**Comment by TheIncorrigible1** [comment]
@vexx32 *cough* `-in`, `-notin`, `@().Where()`, `@().ForEach()`, just to name a few that make my life easier and code more readable.

**Comment by vexx32** [comment]
If it's not allowed, it's time to read up. PS v2 is such a security risk that it's a commonly-tested exploit vector for machines that _have_ 5.1 installed and haven't fully removed v2 yet. Lots of malicious PS scripts explicitly try to use PSv2 because it's harder to protect against attacks from there, not to mention actually properly manage. If memory serves, that was before the real logging solutions were implemented, so most of them can do whatever they want without being particularly traceable.

It's not a good policy, not a _safe_ policy by any measure of the imagination, to retain PSv2 on any company devices. I'm fighting a similar fight at my location, though in my case I'm just glad these old machines are dying off quickly enough and our replacements are _all_ windows 10. 😄

**Comment by DarthSpock** [comment]
I was able to upgrade but my $PSModulePath is fried. I'd hate to revert, save it somewhere, and then upgrade again. Any way to get it back?

**Comment by rjmholt** [comment]
Hi all, just want to let you know that I've been doing some work to help with the compatibility story here:

https://github.com/PowerShell/PSScriptAnalyzer/pull/1133

The next release of PSScriptAnalyzer will (with the right configuration) be able to determine that a command, .NET type or syntax in your script is not available by default on another platform (like PowerShell 3/4). So writing a script in PowerShell 5.1 or 6, you'll get advance warning that `[System.Collections.ArrayList]::new()` or `Get-FileHash $myFile` or `"Stuff" | Out-File ./file.txt -NoNewline` won't work in PowerShell 3/4.

**Comment by rjmholt** [comment]
FYI everyone, I've just published [the blog post](https://devblogs.microsoft.com/powershell/using-psscriptanalyzer-to-check-powershell-version-compatibility/).

**Comment by SydneyhSmith** [comment]
In the next release of the PowerShell extension we will be dropping support for v3/v4. For more information: 
https://devblogs.microsoft.com/powershell/powershell-editor-services-roadmap/
